From a485516e3b875ec7f35f1dcd1427962bcd8a7623 Mon Sep 17 00:00:00 2001
From: Reagan Bohan <xbjfk.github@gmail.com>
Date: Sat, 3 May 2025 10:54:10 +0000
Subject: [PATCH 32/80] Add musl_v1_2_3 feature

This feature, controlled by the environment variable
RUST_LIBC_UNSTABLE_MUSL_V1_2_3 will control whether breaking changes up
to musl v1.2.3 will be reflected

(backport <https://github.com/rust-lang/libc/pull/4443>)
(cherry picked from commit 2a68f7f9f6139f8930df345ae19697336908e940)
---
 build.rs           | 8 ++++++++
 libc-test/build.rs | 3 +++
 2 files changed, 11 insertions(+)

diff --git a/build.rs b/build.rs
index bcb9495e4..2a55a4c07 100644
--- a/build.rs
+++ b/build.rs
@@ -22,6 +22,7 @@ const ALLOWED_CFGS: &[&str] = &[
     "libc_ctest",
     // Corresponds to `__USE_TIME_BITS64` in UAPI
     "linux_time_bits64",
+    "musl_v1_2_3"
 ];
 
 // Extra values to allow for check-cfg.
@@ -89,6 +90,13 @@ fn main() {
         _ => (),
     }
 
+    let musl_v1_2_3 = env::var("RUST_LIBC_UNSTABLE_MUSL_V1_2_3").is_ok();
+    println!("cargo:rerun-if-env-changed=RUST_LIBC_UNSTABLE_MUSL_V1_2_3");
+    // loongarch64 and ohos have already updated
+    if musl_v1_2_3 || target_os == "loongarch64" || target_env == "ohos" {
+        // FIXME(musl): enable time64 api as well
+        set_cfg("musl_v1_2_3");
+    }
     let linux_time_bits64 = env::var("RUST_LIBC_UNSTABLE_LINUX_TIME_BITS64").is_ok();
     println!("cargo:rerun-if-env-changed=RUST_LIBC_UNSTABLE_LINUX_TIME_BITS64");
     if linux_time_bits64 {
diff --git a/libc-test/build.rs b/libc-test/build.rs
index 71758f464..e75f10060 100644
--- a/libc-test/build.rs
+++ b/libc-test/build.rs
@@ -3734,6 +3734,9 @@ fn test_linux(target: &str) {
     let old_musl = musl && !musl_v1_2_3;
 
     let mut cfg = ctest_cfg();
+    if musl_v1_2_3 {
+        cfg.cfg("musl_v1_2_3", None);
+    }
     cfg.define("_GNU_SOURCE", None);
     // This macro re-defines fscanf,scanf,sscanf to link to the symbols that are
     // deprecated since glibc >= 2.29. This allows Rust binaries to link against
-- 
2.43.0

